pkg/controller/keycloakclient/keycloakclient_controller.go

```diff
@@ -0,0 +1,306 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, _ chan schema.GroupVersionKind) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &ReconcileKeycloakClient{
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		cancel:   cancel,
+		context:  ctx,
+		recorder: mgr.GetRecorder(ControllerName),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KeycloakClient
+	err = c.Watch(&source.Kind{Type: &kc.KeycloakClient{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Make sure to watch the credential secrets
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &kc.KeycloakClient{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKeycloakClient implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKeycloakClient{}
+
+// ReconcileKeycloakClient reconciles a KeycloakClient object
+type ReconcileKeycloakClient struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client   client.Client
+	scheme   *runtime.Scheme
+	context  context.Context
+	cancel   context.CancelFunc
+	recorder record.EventRecorder
+}
+
+// Reconcile reads that state of the cluster for a KeycloakClient object and makes changes based on the state read
+// and what is in the KeycloakClient.Spec
+func (r *ReconcileKeycloakClient) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeycloakClient")
+
+	// Fetch the KeycloakClient instance
+	instance := &kc.KeycloakClient{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	err = r.validate(instance)
+	if err != nil {
+		return r.ManageError(instance, err)
+	}
+
+	r.adjustCR(instance)
+
+	keycloaks, err := r.getMatchingKeycloaks(instance)
```

Comment:
This could be switched to just the keycloak in the same namespace as the client creation or even to default to that if no `instanceSelector` is specified in the Spec

Reply:
I agree to that. I think we could always assume that whenever `instanceSelector` is `nil`, we target the same namespace. However, the same mechanics needs to be used for `KeycloakRealm`, `KeycloakClient` and `KeycloakUser`. And since 2 out of 3 are still in a PR state, I believe we should implement it as soon as all the code is merged. Does it makes sense? Otherwise I worry we'll end up with 3 different solutions to the same problem.

Reply:
ya, makes perfect sense. No problem with this being in a separate PR